### PR TITLE
Addition of threading

### PR DESCRIPTION
**Pull request by TrystanLea**
Associated with https://github.com/emonhub/emonhub/issues/66
These are the changes required to add threading. Il try and write it up next

### CODE REVIEW

**Comment by TrystanLea** [comment]
Initial testing shows no error's, I will continue longer term testing.

File: src/emonhub_dispatcher.py [hunk 1]
```diff
@@ -108,11 +115,30 @@ def add(self, data):
...
 
         """
-        self.flush()
+        #lastposttime = time.time()
+        while not self.stop:
+        
+            # Fetch values from queue
+            
+            # Send every 5 seconds
+            #if ((time.time() - lastposttime) > 5.0):
```
Comment (TrystanLea) [inline src/emonhub_dispatcher.py]: The if ((time.time() - lastposttime) > 5.0): part here was an option to create a minimum post time that data could be dispatched but perhaps something to think about in the future.

**Comment by pb66** [comment]
Thanks @TrystanLea some notes would be helpful, although I am getting a gist of what's what from the code. 
If I'm reading it right there maybe a small change required to when the [dispatchers are started](https://github.com/TrystanLea/emonhub/blob/development/src/emonhub.py#L79-L81). If a dispatcher is added after the initial start up, it looks like it will get [created with a queue ok,](https://github.com/TrystanLea/emonhub/blob/development/src/emonhub.py#L160-L162) but will not get started, 
Any reason why the dispatcher cannot be started with self.start() as the last line in the dispatchers **init**().
The ability to delete and create dispatchers whilst running will effect the ability to edit the init_settings without restarting emonHub.
I will give this a run and come back to you, but it looks good.

**Comment by TrystanLea** [comment]
Great, thanks Paul. Did you try putting the start in self.start? Sounds like it might work?

**Comment by pb66** [comment]
Yes I did and it appears to work ok, emonHub certainly seems much more responsive with threading and as a by product the logs are a little tidier as the "appends" all happen before all the "sends" etc

```
2014-07-20 09:59:48,444 INFO  NEW FRAME : 1405846788.44 18 1.0 0
2014-07-20 09:59:48,444 DEBUG Timestamp : 1405846788.44
2014-07-20 09:59:48,444 DEBUG      Node : 18
2014-07-20 09:59:48,444 DEBUG    Values : [1, 0]
2014-07-20 09:59:48,536 DEBUG Append to 'emonCMS_local' buffer => time: 1405846788.44, data: [18, 1, 0]
2014-07-20 09:59:48,536 DEBUG Append to 'emonCMS' buffer => time: 1405846788.44, data: [18, 1, 0]
2014-07-20 09:59:48,635 INFO Sending: http://emoncms.org/input/bulk.json?apikey=E-M-O-N-C-M-S-A-P-I-K-E-Y&data=[[1405846788.44,18,1,0]]&sentat=1405846788
2014-07-20 09:59:48,635 INFO Sending: http://192.168.0.214/emoncms/input/bulk.json?apikey=E-M-O-N-C-M-S-A-P-I-K-E-Y&data=[[1405846788.44,18,1,0]]&sentat=1405846788
2014-07-20 09:59:48,901 DEBUG Receipt acknowledged with 'ok' from http://emoncms.org
2014-07-20 09:59:49,385 DEBUG Receipt acknowledged with 'ok' from http://192.168.0.214/emoncms
```

I have also resolved a little issue with the pause functions and I'm currently trying the existing interval timer, I will post what I've got shortly, for you to check over if you wouldn't mind.

**Comment by pb66** [comment]
I have created a multi-threading branch to experiment with threading #70

**Comment by TrystanLea** [comment]
I've updated the bare bones threading example I created a while ago while learning about threading here: https://github.com/emoncms/development/blob/master/experimental/emonhub_development/PythonThreading/01.py

There are 3 main parts to adding threading:

#### 1) Defining the dispatcher class as a thread

```
import threading
class Dispatcher(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self)
    def run(self):
        # do something here
```

The 'thread' class is then started from the main program with dispatcher.start()

#### 2) Adding a mechanism to close the thread when you try and close the program (ie via ctrl+c)

We can do this by defining a property in the Dispatcher class that can be switched from the main program in order to signal that the dispatcher thread needs to exit. (See: [http://www.regexprn.com/2010/05/killing-multithreaded-python-programs.html](http://www.regexprn.com/2010/05/killing-multithreaded-python-programs.html) for more on this)

In the main program loop we have: (in _def _sigint_handler_ and _def close_ in emonhub)

```
except KeyboardInterrupt:
    dispatcher.stop = True
    break
```

Our dispatcher then looks like this:

```
class Dispatcher(threading.Thread):
    def __init__(self):
        threading.Thread.__init__(self)
        self.stop = False
    def run(self):
        while not self.stop:
            # thread loop 
```

#### 3) Add a queue to pass data to be dispatched from the main program to the dispatcher thread

We can create a queue in the main program with:

```
queue = Queue.Queue(0)
```

The queue object needs to be passed to the dispatcher as a variable in the dispatcher init method when it is created, and then its just a matter of calling

```
queue.put(i)
```

in the main program to put data into the queue, and then to fetch the data from the queue in the dispatcher:

```
if not self._queue.empty():
    item = self._queue.get()
```

I think that covers the main parts, let me know if I can clarify any particular part.

**Comment by pb66** [comment]
Thanks for this it help alot

**Comment by TrystanLea** [comment]
Great, thanks Paul, Il switch my test systems to the multi-threading branch to test. What do you think are the steps before merging it into the master branch?

**Comment by pb66** [comment]
I've been using the "development" branch as the "master" in Jerome's and Dave's absence, a bit of testing and I am happy to merge "multi-threading" into "development". 

I will update the "v1 release proposal" issue #35 soon, as I would really like to move towards a v1 rollout.

**Comment by pb66** [comment]
I will close this issue for now and link to these notes from #66

**Comment by TrystanLea** [comment]
Great, thanks Paul

**Pull request closed without merging**